src/core/qgspointlocator.cpp

```diff
@@ -332,7 +332,16 @@ static QgsPointLocator::MatchList _geometrySegmentsInRect( QgsGeometry *geom, co
   // we need iterator for segments...
 
   QgsPointLocator::MatchList lst;
-  QByteArray wkb( geom->asWkb() );
+
+  QgsGeometry straightGeom = geom->convertToType( QgsWkbTypes::LineGeometry, true );
+  straightGeom.convertToStraightSegment();
+
+  QgsWkbTypes::Type wkbType = straightGeom.wkbType();
+
+  if ( ( straightGeom.type() != QgsWkbTypes::LineGeometry ) && ( !straightGeom.isMultipart() ) )
```

Comment:
I'm still struggling to understand this one -- we are allowing ANY line types (single or multi), and ANY multi part types (multipoint, multipolygon, etc). Why multipolygon/multipoint but NOT polygon/point? 

I'd suggest if we could rip out the wkb parsing here and use the QgsAbstractGeometry classes directly it'd make everything much easier to understand (and also faster!)

Reply:
The points are not to be taken in this case, since they are intersections. We can ask ourselves the question of the case where a point is on a line, but it is a snapping on a vertex normally and not an intersection. It was the logic already present that I think is the right one and that I have kept.

For polygons, they are converted first. A polygon becomes a multilinestring. As I mentioned, it was a suggestion to avoid having two cases in the intersection algorithm after: multilinestring and linestring.

The function retrieves the geometry, converts it into multilinestring `straightGeom = geom->convertToType( QgsWkbTypes::LineGeometry, true )` and segments the curves if necessary `straightGeom.convertToStraightSegment()`.

So you can only have multilinestring.

Then we use the previous code on multilinestring to apply it to geometries.

Maybe I should comment on the code to explain this.

Reply:
@nyalldawson is my explanation sufficient? Do you want me to add a comment in the code? Thanks

Reply:
So in case of multipoint, the code in here will be executed?